File: backend/app/rag/structured_answer.py
from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum

from app.rag.resolution import CitationResolutionStatus, ResolvedAnswerDraft, ResolvedPlaceholder
from app.rag.self_rag import SelfRAGClaimResult, SelfRAGClaimStatus, SelfRAGVerificationResult


class CitationBadgeStatus(StrEnum):
    VERIFIED = "VERIFIED"
    UNCERTAIN = "UNCERTAIN"
    UNVERIFIED = "UNVERIFIED"


class StructuredAnswerSectionKind(StrEnum):
    LEGAL_POSITION = "LEGAL_POSITION"
    APPLICABLE_LAW = "APPLICABLE_LAW"
    KEY_CASES = "KEY_CASES"
    VERIFICATION_STATUS = "VERIFICATION_STATUS"


@dataclass(slots=True, frozen=True)
class InlineCitationBadge:
    placeholder_token: str
    label: str
    status: CitationBadgeStatus
    citation: str | None
    message: str
    doc_id: str | None
    chunk_id: str | None
    source_passage: str | None
    appeal_warning: str | None


@dataclass(slots=True, frozen=True)
class StructuredClaim:
    text: str
    status: CitationBadgeStatus
    reason: str
    citation: str | None
    source_passage: str | None
    appeal_warning: str | None
    reretrieved: bool
    citation_badges: tuple[InlineCitationBadge, ...]


@dataclass(slots=True, frozen=True)
class VerificationStatusItem:
    label: str
    value: str
    status: CitationBadgeStatus


@dataclass(slots=True, frozen=True)
class StructuredAnswerSection:
    kind: StructuredAnswerSectionKind
    title: str
    claims: tuple[StructuredClaim, ...] = ()
    status_items: tuple[VerificationStatusItem, ...] = ()


@dataclass(slots=True, frozen=True)
class StructuredAnswer:
    query: str
    overall_status: CitationBadgeStatus
    sections: tuple[StructuredAnswerSection, ...]

    def section(
        self,
        kind: StructuredAnswerSectionKind,
    ) -> StructuredAnswerSection:
        for section in self.sections:
            if section.kind is kind:
                return section
        raise KeyError(f"Section {kind} not found")


class StructuredAnswerBuilder:
    def build(
        self,
        *,
        resolved_draft: ResolvedAnswerDraft,
        verification_result: SelfRAGVerificationResult,
    ) -> StructuredAnswer:
        grouped_claims: dict[StructuredAnswerSectionKind, list[StructuredClaim]] = {
            StructuredAnswerSectionKind.LEGAL_POSITION: [],
            StructuredAnswerSectionKind.APPLICABLE_LAW: [],
            StructuredAnswerSectionKind.KEY_CASES: [],
        }

        for claim in verification_result.claims:
            section_kind = self._map_section_kind(claim.section_title)
            grouped_claims[section_kind].append(
                self._build_claim(
                    claim=claim,
                    resolved_draft=resolved_draft,
                )
            )

        verification_section = StructuredAnswerSection(
            kind=StructuredAnswerSectionKind.VERIFICATION_STATUS,
            title="Verification Status",
            status_items=self._build_verification_items(
                resolved_draft=resolved_draft,
                verification_result=verification_result,
            ),
        )

        sections = (
            StructuredAnswerSection(
                kind=StructuredAnswerSectionKind.LEGAL_POSITION,
                title="Legal Position",
                claims=tuple(grouped_claims[StructuredAnswerSectionKind.LEGAL_POSITION]),
            ),
            StructuredAnswerSection(
                kind=StructuredAnswerSectionKind.APPLICABLE_LAW,
                title="Applicable Law",
                claims=tuple(grouped_claims[StructuredAnswerSectionKind.APPLICABLE_LAW]),
            ),
            StructuredAnswerSection(
                kind=StructuredAnswerSectionKind.KEY_CASES,
                title="Key Cases",
                claims=tuple(grouped_claims[StructuredAnswerSectionKind.KEY_CASES]),
            ),
            verification_section,
        )

        return StructuredAnswer(
            query=resolved_draft.draft.query,
            overall_status=self._overall_status(verification_result),
            sections=sections,
        )

    def _build_claim(
        self,
        *,
        claim: SelfRAGClaimResult,
        resolved_draft: ResolvedAnswerDraft,
    ) -> StructuredClaim:
        badges = tuple(
            self._build_badge(
                claim=claim,
                resolution=resolved_draft.resolution_for(token),
                placeholder_token=token,
            )
            for token in claim.placeholder_tokens
        )

        if not badges and claim.citation is not None:
            badges = (
                InlineCitationBadge(
                    placeholder_token="",
                    label=claim.citation,
                    status=self._status_for_claim(claim.status),
                    citation=claim.citation,
                    message=claim.reason,
                    doc_id=claim.source_doc_id,
                    chunk_id=claim.source_chunk_id,
                    source_passage=claim.source_passage,
                    appeal_warning=claim.appeal_warning,
                ),
            )

        return StructuredClaim(
            text=claim.claim,
            status=self._status_for_claim(claim.status),
            reason=claim.reason,
            citation=claim.citation,
            source_passage=claim.source_passage,
            appeal_warning=claim.appeal_warning,
            reretrieved=claim.reretrieved,
            citation_badges=badges,
        )

    def _build_badge(
        self,
        *,
        claim: SelfRAGClaimResult,
        resolution: ResolvedPlaceholder | None,
        placeholder_token: str,
    ) -> InlineCitationBadge:
        status = self._status_for_resolution(claim=claim, resolution=resolution)
        label = self._badge_label(claim=claim, resolution=resolution)
        message_parts = [claim.reason]
        if resolution is not None and resolution.message not in message_parts:
            message_parts.append(resolution.message)

        return InlineCitationBadge(
            placeholder_token=placeholder_token,
            label=label,
            status=status,
            citation=claim.citation if status is not CitationBadgeStatus.UNVERIFIED else None,
            message=" ".join(part for part in message_parts if part),
            doc_id=claim.source_doc_id or (resolution.doc_id if resolution is not None else None),
            chunk_id=claim.source_chunk_id
            or (resolution.chunk_id if resolution is not None else None),
            source_passage=claim.source_passage,
            appeal_warning=claim.appeal_warning,
        )

    def _build_verification_items(
        self,
        *,
        resolved_draft: ResolvedAnswerDraft,
        verification_result: SelfRAGVerificationResult,
    ) -> tuple[VerificationStatusItem, ...]:
        verified_resolutions = sum(
            1
            for resolution in resolved_draft.resolutions
            if resolution.status is CitationResolutionStatus.VERIFIED
        )
        unverified_resolutions = len(resolved_draft.resolutions) - verified_resolutions

        return (
            VerificationStatusItem(
                label="Verified Claims",
                value=str(verification_result.verified_count),
                status=CitationBadgeStatus.VERIFIED,
            ),
            VerificationStatusItem(
                label="Claims Requiring Review",
                value=str(verification_result.uncertain_count),
                status=(
                    CitationBadgeStatus.UNCERTAIN
                    if verification_result.uncertain_count > 0
                    else CitationBadgeStatus.VERIFIED
                ),
            ),
            VerificationStatusItem(
                label="Unverified Claims",
                value=str(verification_result.unsupported_count),
                status=(
                    CitationBadgeStatus.UNVERIFIED
                    if verification_result.unsupported_count > 0
                    else CitationBadgeStatus.VERIFIED
                ),
            ),
            VerificationStatusItem(
                label="Resolved Citations",
                value=str(verified_resolutions),
                status=CitationBadgeStatus.VERIFIED,
            ),
            VerificationStatusItem(
                label="Unresolved Citations",
                value=str(unverified_resolutions),
                status=(
                    CitationBadgeStatus.UNVERIFIED
                    if unverified_resolutions > 0
                    else CitationBadgeStatus.VERIFIED
                ),
            ),
        )

    def _overall_status(
        self,
        verification_result: SelfRAGVerificationResult,
    ) -> CitationBadgeStatus:
        if verification_result.unsupported_count > 0:
            return CitationBadgeStatus.UNVERIFIED
        if verification_result.uncertain_count > 0:
            return CitationBadgeStatus.UNCERTAIN
        return CitationBadgeStatus.VERIFIED

    def _map_section_kind(self, title: str) -> StructuredAnswerSectionKind:
        normalized = title.strip().lower()
        if "law" in normalized or "statute" in normalized:
            return StructuredAnswerSectionKind.APPLICABLE_LAW
        if "authorit" in normalized or "case" in normalized:
            return StructuredAnswerSectionKind.KEY_CASES
        return StructuredAnswerSectionKind.LEGAL_POSITION

    def _status_for_claim(
        self,
        status: SelfRAGClaimStatus,
    ) -> CitationBadgeStatus:
        if status is SelfRAGClaimStatus.VERIFIED:
            return CitationBadgeStatus.VERIFIED
        if status is SelfRAGClaimStatus.UNCERTAIN:
            return CitationBadgeStatus.UNCERTAIN
        return CitationBadgeStatus.UNVERIFIED

    def _status_for_resolution(
        self,
        *,
        claim: SelfRAGClaimResult,
        resolution: ResolvedPlaceholder | None,
    ) -> CitationBadgeStatus:
        if resolution is None or resolution.status is CitationResolutionStatus.UNVERIFIED:
            return CitationBadgeStatus.UNVERIFIED
        return self._status_for_claim(claim.status)

    def _badge_label(
        self,
        *,
        claim: SelfRAGClaimResult,
        resolution: ResolvedPlaceholder | None,
    ) -> str:
        if resolution is None:
            return claim.citation or "[UNVERIFIED]"
        if resolution.status is CitationResolutionStatus.UNVERIFIED:
            return resolution.rendered_value
        if claim.citation is not None and len(claim.placeholder_tokens) == 1:
            return claim.citation
        return resolution.rendered_value

File: frontend/lib/structured-answer.ts
export type StructuredAnswerBadgeStatus =
  | "VERIFIED"
  | "UNCERTAIN"
  | "UNVERIFIED";

export type StructuredAnswerSectionKind =
  | "LEGAL_POSITION"
  | "APPLICABLE_LAW"
  | "KEY_CASES"
  | "VERIFICATION_STATUS";

export type InlineCitationBadge = {
  appealWarning: string | null;
  chunkId: string | null;
  citation: string | null;
  docId: string | null;
  label: string;
  message: string;
  placeholderToken: string;
  sourcePassage: string | null;
  status: StructuredAnswerBadgeStatus;
};

export type StructuredClaim = {
  appealWarning: string | null;
  citation: string | null;
  citationBadges: InlineCitationBadge[];
  reason: string;
  reretrieved: boolean;
  sourcePassage: string | null;
  status: StructuredAnswerBadgeStatus;
  text: string;
};

export type VerificationStatusItem = {
  label: string;
  status: StructuredAnswerBadgeStatus;
  value: string;
};

export type StructuredAnswerSection = {
  kind: StructuredAnswerSectionKind;
  title: string;
  claims: StructuredClaim[];
  statusItems: VerificationStatusItem[];
};

export type StructuredAnswer = {
  overallStatus: StructuredAnswerBadgeStatus;
  query: string;
  sections: StructuredAnswerSection[];
};

export type StructuredAnswerSource = {
  appealWarning: string | null;
  chunkId: string | null;
  citation: string;
  docId: string | null;
  id: string;
  label: string;
  message: string;
  sourcePassage: string | null;
  status: StructuredAnswerBadgeStatus;
};

export function buildStructuredSourceId(input: {
  chunkId: string | null;
  citation: string | null;
  docId: string | null;
  label: string;
}): string {
  return [
    input.docId ?? "",
    input.chunkId ?? "",
    input.citation ?? input.label,
  ].join("::");
}

function sourceIdForBadge(badge: InlineCitationBadge): string {
  return buildStructuredSourceId({
    docId: badge.docId,
    chunkId: badge.chunkId,
    citation: badge.citation,
    label: badge.label,
  });
}

export function collectStructuredAnswerSources(
  answer: StructuredAnswer,
): StructuredAnswerSource[] {
  const seen = new Set<string>();
  const sources: StructuredAnswerSource[] = [];

  for (const section of answer.sections) {
    for (const claim of section.claims) {
      for (const badge of claim.citationBadges) {
        if (!badge.citation && !badge.sourcePassage) {
          continue;
        }

        const id = sourceIdForBadge(badge);
        if (seen.has(id)) {
          continue;
        }

        seen.add(id);
        sources.push({
          id,
          label: badge.label,
          citation: badge.citation ?? badge.label,
          status: badge.status,
          message: badge.message,
          docId: badge.docId,
          chunkId: badge.chunkId,
          sourcePassage: badge.sourcePassage,
          appealWarning: badge.appealWarning,
        });
      }
    }
  }

  return sources;
}

export const demoStructuredAnswer: StructuredAnswer = {
  query:
    "What are the strongest anticipatory bail arguments on these facts, and which binding Supreme Court cases should appear first in the note?",
  overallStatus: "UNCERTAIN",
  sections: [
    {
      kind: "LEGAL_POSITION",
      title: "Legal Position",
      statusItems: [],
      claims: [
        {
          text: "On the present record, the strongest anticipatory bail submission is that the prosecution narrative still looks substantially documentary and commercial, which weakens any immediate custodial justification.",
          status: "VERIFIED",
          reason:
            "Binding anticipatory bail authorities emphasize concrete necessity for custody, not generalized investigation language.",
          citation: "Siddharam Satlingappa Mhetre v State of Maharashtra, (2011) 1 SCC 694",
          sourcePassage:
            "Personal liberty requires the court to examine whether the prosecution has shown concrete investigative necessity before permitting arrest in anticipatory bail matters.",
          appealWarning: null,
          reretrieved: false,
          citationBadges: [
            {
              placeholderToken: "[CITE: anticipatory bail liberty authority]",
              label: "Siddharam Satlingappa Mhetre",
              status: "VERIFIED",
              citation:
                "Siddharam Satlingappa Mhetre v State of Maharashtra, (2011) 1 SCC 694",
              message:
                "Verified authority supporting a liberty-first anticipatory bail analysis.",
              docId: "sc-siddharam-2011",
              chunkId: "chunk-3",
              sourcePassage:
                "Personal liberty requires the court to examine whether the prosecution has shown concrete investigative necessity before permitting arrest in anticipatory bail matters.",
              appealWarning: null,
            },
            {
              placeholderToken: "[CITE: foundational anticipatory bail bench]",
              label: "Gurbaksh Singh Sibbia",
              status: "VERIFIED",
              citation:
                "Gurbaksh Singh Sibbia v State of Punjab, (1980) 2 SCC 565",
              message:
                "Constitution Bench baseline on anticipatory bail discretion.",
              docId: "sc-sibbia-1980",
              chunkId: "chunk-8",
              sourcePassage:
                "Anticipatory bail must remain flexible and fact-sensitive, but it cannot be denied merely on broad prosecutorial assertions.",
              appealWarning: null,
            },
          ],
        },
        {
          text: "The prosecution will still argue that financial-trail recovery and witness coordination demand custodial access, so the defense must foreground cooperation and document production.",
          status: "UNCERTAIN",
          reason:
            "The cited arrest-guideline authority is relevant, but factual parity depends on the prosecution's current material.",
          citation: "Arnesh Kumar v State of Bihar, (2014) 8 SCC 273",
          sourcePassage:
            "Arrest cannot be routine, and the investigating officer must justify why custody is necessary on the facts of the case.",
          appealWarning: null,
          reretrieved: true,
          citationBadges: [
            {
              placeholderToken: "[CITE: arrest restraint authority]",
              label: "Arnesh Kumar",
              status: "UNCERTAIN",
              citation: "Arnesh Kumar v State of Bihar, (2014) 8 SCC 273",
              message:
                "Useful arrest-restraint authority, but still needs fact-specific application to this record.",
              docId: "sc-arnesh-2014",
              chunkId: "chunk-2",
              sourcePassage:
                "Arrest cannot be routine, and the investigating officer must justify why custody is necessary on the facts of the case.",
              appealWarning: null,
            },
          ],
        },
      ],
    },
    {
      kind: "APPLICABLE_LAW",
      title: "Applicable Law",
      statusItems: [],
      claims: [
        {
          text: "For a post-July 2024 analysis, the anticipatory bail reference point is BNSS Section 482, even if the FIR and draft pleadings still use the older CrPC 438 vocabulary.",
          status: "VERIFIED",
          reason:
            "The workspace already maps legacy criminal procedure references to the new criminal code regime.",
          citation: "BNSS Section 482",
          sourcePassage:
            "BNSS Section 482 carries forward the anticipatory bail framework for the post-cutover criminal procedure regime.",
          appealWarning: null,
          reretrieved: false,
          citationBadges: [
            {
              placeholderToken: "[STATUTE: BNSS, Section 482]",
              label: "BNSS 482",
              status: "VERIFIED",
              citation: "Bharatiya Nagarik Suraksha Sanhita, 2023, Section 482",
              message: "Current anticipatory bail provision for post-cutover analysis.",
              docId: "bnss-482",
              chunkId: "section-482",
              sourcePassage:
                "BNSS Section 482 carries forward the anticipatory bail framework for the post-cutover criminal procedure regime.",
              appealWarning: null,
            },
          ],
        },
        {
          text: "The cheating and breach-of-trust allegations also need to be tracked against BNS Sections 318 and 316 so the note stays temporally current.",
          status: "VERIFIED",
          reason:
            "Temporal validity and code mapping are part of the system's statutory verification path.",
          citation: "BNS Sections 318 and 316",
          sourcePassage:
            "The new penal code provisions preserve the operative cheating and breach-of-trust categories while changing section numbering.",
          appealWarning: null,
          reretrieved: false,
          citationBadges: [
            {
              placeholderToken: "[STATUTE: BNS, Section 318]",
              label: "BNS 318",
              status: "VERIFIED",
              citation: "Bharatiya Nyaya Sanhita, 2023, Section 318",
              message: "Mapped from IPC 420 for post-cutover treatment.",
              docId: "bns-318",
              chunkId: "section-318",
              sourcePassage:
                "The new penal code provisions preserve the operative cheating categories while changing section numbering.",
              appealWarning: null,
            },
            {
              placeholderToken: "[STATUTE: BNS, Section 316]",
              label: "BNS 316",
              status: "VERIFIED",
              citation: "Bharatiya Nyaya Sanhita, 2023, Section 316",
              message: "Mapped from IPC 406 for post-cutover treatment.",
              docId: "bns-316",
              chunkId: "section-316",
              sourcePassage:
                "The new penal code provisions preserve the operative breach-of-trust categories while changing section numbering.",
              appealWarning: null,
            },
          ],
        },
      ],
    },
    {
      kind: "KEY_CASES",
      title: "Key Cases",
      statusItems: [],
      claims: [
        {
          text: "Siddharam Satlingappa Mhetre should lead the note because it gives the cleanest liberty-first framework for anticipatory bail where the State has not yet justified custody with precision.",
          status: "VERIFIED",
          reason: "Best binding fit for the current bail posture.",
          citation:
            "Siddharam Satlingappa Mhetre v State of Maharashtra, (2011) 1 SCC 694",
          sourcePassage:
            "Personal liberty requires the court to examine whether the prosecution has shown concrete investigative necessity before permitting arrest in anticipatory bail matters.",
          appealWarning: null,
          reretrieved: false,
          citationBadges: [
            {
              placeholderToken: "[CITE: lead anticipatory bail authority]",
              label: "Siddharam Satlingappa Mhetre",
              status: "VERIFIED",
              citation:
                "Siddharam Satlingappa Mhetre v State of Maharashtra, (2011) 1 SCC 694",
              message: "Lead binding authority for this answer.",
              docId: "sc-siddharam-2011",
              chunkId: "chunk-3",
              sourcePassage:
                "Personal liberty requires the court to examine whether the prosecution has shown concrete investigative necessity before permitting arrest in anticipatory bail matters.",
              appealWarning: null,
            },
          ],
        },
        {
          text: "Gurbaksh Singh Sibbia remains the foundational Constitution Bench authority on how anticipatory bail discretion should be exercised without reducing the remedy to a rarity.",
          status: "VERIFIED",
          reason: "Foundational binding authority with continuing precedential weight.",
          citation: "Gurbaksh Singh Sibbia v State of Punjab, (1980) 2 SCC 565",
          sourcePassage:
            "Anticipatory bail must remain flexible and fact-sensitive, but it cannot be denied merely on broad prosecutorial assertions.",
          appealWarning: null,
          reretrieved: false,
          citationBadges: [
            {
              placeholderToken: "[CITE: foundational anticipatory bail case]",
              label: "Gurbaksh Singh Sibbia",
              status: "VERIFIED",
              citation:
                "Gurbaksh Singh Sibbia v State of Punjab, (1980) 2 SCC 565",
              message: "Foundational Constitution Bench authority.",
              docId: "sc-sibbia-1980",
              chunkId: "chunk-8",
              sourcePassage:
                "Anticipatory bail must remain flexible and fact-sensitive, but it cannot be denied merely on broad prosecutorial assertions.",
              appealWarning: null,
            },
          ],
        },
      ],
    },
    {
      kind: "VERIFICATION_STATUS",
      title: "Verification Status",
      claims: [],
      statusItems: [
        {
          label: "Verified Claims",
          value: "5",
          status: "VERIFIED",
        },
        {
          label: "Claims Requiring Review",
          value: "1",
          status: "UNCERTAIN",
        },
        {
          label: "Unverified Claims",
          value: "0",
          status: "VERIFIED",
        },
        {
          label: "Resolved Citations",
          value: "6",
          status: "VERIFIED",
        },
        {
          label: "Unresolved Citations",
          value: "0",
          status: "VERIFIED",
        },
      ],
    },
  ],
};
